File: ragtoolkit/sdk/connectors/anthropic.py
"""
Anthropic Connector for RAG Toolkit

Wraps Anthropic client with automatic tracing capabilities.
"""

import functools
import time
from typing import Any, Dict, Optional, Union, List
from ..tracer import RAGTracker


def with_tracing(func):
    """Decorator to add tracing to Anthropic API calls"""
    @functools.wraps(func)
    def wrapper(self, *args, **kwargs):
        # Extract common parameters
        messages = kwargs.get('messages', args[0] if args else [])
        model = kwargs.get('model', getattr(self, 'model', 'claude-3-sonnet-20240229'))
        
        # Start tracing
        tracker = RAGTracker()
        
        # Extract user input from messages
        user_input = None
        if messages and isinstance(messages, list):
            user_messages = [m for m in messages if m.get('role') == 'user']
            if user_messages:
                user_input = user_messages[-1].get('content', '')
        
        with tracker.trace_context(user_input=user_input or 'Anthropic API call') as trace:
            # Add system/user prompts
            if messages:
                full_prompt = "\n".join([f"{m.get('role', 'unknown')}: {m.get('content', '')}" 
                                       for m in messages])
                tracker.add_prompt(full_prompt)
            
            # Add system prompt if provided separately
            system_prompt = kwargs.get('system')
            if system_prompt:
                tracker.add_prompt(f"system: {system_prompt}")
            
            # Record start time
            start_time = time.time()
            
            try:
                # Make the original API call
                response = func(self, *args, **kwargs)
                
                # Record timing
                end_time = time.time()
                latency_ms = int((end_time - start_time) * 1000)
                
                # Extract response content and tokens
                if hasattr(response, 'content') and response.content:
                    # Anthropic returns content as a list of content blocks
                    if isinstance(response.content, list) and response.content:
                        content = response.content[0].text if hasattr(response.content[0], 'text') else str(response.content[0])
                    else:
                        content = str(response.content)
                    
                    tracker.set_model_output(
                        output=content,
                        model_name=model,
                        tokens_in=getattr(response.usage, 'input_tokens', None) if hasattr(response, 'usage') else None,
                        tokens_out=getattr(response.usage, 'output_tokens', None) if hasattr(response, 'usage') else None,
                        latency_ms=latency_ms
                    )
                
                return response
                
            except Exception as e:
                # Record error
                tracker.set_model_output(
                    output=f"Error: {str(e)}",
                    model_name=model,
                    latency_ms=int((time.time() - start_time) * 1000)
                )
                raise
    
    return wrapper


class Complete:
    """
    Drop-in replacement for Anthropic Messages with automatic tracing.
    
    Usage:
        from ragtoolkit.connectors.anthropic import Complete
        from anthropic import Anthropic
        
        client = Anthropic()
        # Replace: response = client.messages.create(...)
        response = Complete.create(client, messages=[...], model="claude-3-sonnet-20240229")
    """
    
    @staticmethod
    @with_tracing
    def create(client, **kwargs):
        """Create a message completion with automatic tracing"""
        return client.messages.create(**kwargs)
    
    @classmethod 
    def wrap_client(cls, client):
        """
        Wrap an existing Anthropic client to add tracing to all message completions.
        
        Args:
            client: Anthropic client instance
            
        Returns:
            Modified client with tracing enabled
        """
        original_create = client.messages.create
        client.messages.create = with_tracing(original_create)
        return client


def traced_anthropic_client(api_key: Optional[str] = None, **kwargs):
    """
    Create a new Anthropic client with tracing enabled.
    
    Args:
        api_key: Anthropic API key
        **kwargs: Additional arguments for Anthropic client
        
    Returns:
        Anthropic client with automatic tracing
    """
    try:
        from anthropic import Anthropic
    except ImportError:
        raise ImportError("anthropic package is required. Install with: pip install anthropic")
    
    client = Anthropic(api_key=api_key, **kwargs)
    return Complete.wrap_client(client) 

File: typescript-sdk/src/connectors/anthropic.ts
/**
 * Anthropic connector with automatic RAG tracing - TypeScript SDK
 */

import {
  AnthropicMessage,
  AnthropicCompletionRequest,
  AnthropicCompletionResponse,
  ChunkData
} from '../types';
import { tracer } from '../tracer';

/**
 * Traced Anthropic Messages function
 * Drop-in replacement for anthropic.messages.create()
 */
export async function createMessage(
  params: AnthropicCompletionRequest,
  anthropicClient?: any
): Promise<AnthropicCompletionResponse> {
  const userMessage = params.messages.find((m: AnthropicMessage) => m.role === 'user');
  const userInput = userMessage?.content || 'Anthropic Message';

  return tracer.startTrace(userInput, async (ctx) => {
    const startTime = Date.now();

    try {
      // Add prompt to trace
      let fullPrompt = '';
      if (params.system) {
        fullPrompt += `system: ${params.system}\n`;
      }
      fullPrompt += params.messages.map((m: AnthropicMessage) => `${m.role}: ${m.content}`).join('\n');
      ctx.addPrompt(fullPrompt);

      let response: AnthropicCompletionResponse;

      if (anthropicClient) {
        // Use provided client
        response = await anthropicClient.messages.create(params);
      } else {
        // Try to use global Anthropic client if available
        try {
          // @ts-ignore - Anthropic might be available globally
          const Anthropic = require('@anthropic-ai/sdk');
          const client = new Anthropic();
          response = await client.messages.create(params);
        } catch (error) {
          throw new Error('Anthropic client not provided and global Anthropic not available. Install @anthropic-ai/sdk package or pass client.');
        }
      }

      // Extract response data
      const output = response.content[0]?.text || '';
      const tokensIn = response.usage?.input_tokens;
      const tokensOut = response.usage?.output_tokens;
      const latencyMs = Date.now() - startTime;

      ctx.setModelOutput(output, params.model, tokensIn, tokensOut);

      return response;
    } catch (error) {
      ctx.setError(error instanceof Error ? error.message : String(error));
      throw error;
    }
  }) as Promise<AnthropicCompletionResponse>;
}

/**
 * Wrap an existing Anthropic client with automatic tracing
 */
export function wrapAnthropicClient(client: any): any {
  const originalCreate = client.messages.create.bind(client.messages);

  client.messages.create = async (params: AnthropicCompletionRequest) => {
    return createMessage(params, { messages: { create: originalCreate } });
  };

  return client;
}

/**
 * Create a traced Anthropic client factory
 */
export function createTracedAnthropic(Anthropic: any, options?: any): any {
  const client = new Anthropic(options);
  return wrapAnthropicClient(client);
}

/**
 * RAG-specific utilities for Anthropic
 */
export const anthropicUtils = {
  /**
   * Helper to add retrieved chunks before Anthropic call
   */
  async createMessageWithRAG(
    params: AnthropicCompletionRequest,
    retrievedChunks: ChunkData[],
    scores?: number[],
    anthropicClient?: any
  ): Promise<AnthropicCompletionResponse> {
    const userMessage = params.messages.find((m: AnthropicMessage) => m.role === 'user');
    const userInput = userMessage?.content || 'Anthropic RAG Query';

    return tracer.startTrace(userInput, async (ctx) => {
      // Add retrieved chunks to trace
      ctx.addRetrievedChunks(retrievedChunks, scores);

      // Modify the prompt to include context
      const context = retrievedChunks.map(chunk => chunk.text).join('\n\n');
      const modifiedMessages = [...params.messages];
      
      // Find the last user message and enhance it with context
      const lastUserMsgIndex = modifiedMessages.map(m => m.role).lastIndexOf('user');
      if (lastUserMsgIndex !== -1) {
        modifiedMessages[lastUserMsgIndex] = {
          ...modifiedMessages[lastUserMsgIndex],
          content: `Context:\n${context}\n\nQuestion: ${modifiedMessages[lastUserMsgIndex].content}`
        };
      }

      const modifiedParams = { ...params, messages: modifiedMessages };
      
      // Call the regular traced message creation
      return createMessage(modifiedParams, anthropicClient);
    }) as Promise<AnthropicCompletionResponse>;
  }
};

// Export types for convenience
export type { AnthropicMessage, AnthropicCompletionRequest, AnthropicCompletionResponse }; 